### DISCUSSION TITLE
Queryraw condition joining

### DISCUSSION THREAD

**Discussion by NubeBuster** [prisma/prisma#24838] [Q&A] [answered]
### Question

I need to use a procedure and then filter on this from a sqlserver database. I am having some issues constructing a query that is dynamic and safe. Prisma.join is only for joining values, but how do I join conditions?
What makes it complicated is that the conditions are dynamic filters. Some some may be present or not. 

```ts
const whereClause = []

    if (zipCodeLike && zipCodeLike.length > 0) {
      const clauses = zipCodeLike.map((_, index) =>
        index === 0 ? ` recentZIPCode LIKE ` : ` OR recentZIPCode LIKE `,
      );
      whereClause.push(Prisma.sql([...clauses, ""], ...zipCodeLike));
    }


    if (nameLike && nameLike.length > 0) {
      const clauses = nameLike.map((_, index) =>
        index === 0 ? ` name LIKE ` : ` OR name LIKE `,
      );
      whereClause.push(Prisma.sql([...clauses, ""], ...nameLike));
    }

```

How do I construct the where clause into a queryRaw in a way that they have brackets and AND between them? I tried explaining ChatGPT that the Prisma.join function is only for values but it cannot wrap it's processor around it. Perhaps what I am trying to do is impossible (and hence maybe bad)

## Reproduction
Because I have had this issue in two discords for months without a single response, I've made a minimal reproducible repo playground. https://github.com/Nubebuster/prisma-raw

You can play around in it to find a solution.

To run the test environment run the following commands
> pnpm install
> pnpm prisma generate
> pnpm prisma db push
> pnpm dev

**Comment by NubeBuster** [comment]
Take for example this code:

```ts
    const conditions = [];
    const params = [];

    if (bookingCount !== undefined) {
      conditions.push(`bookingCount > ?`);
      params.push(`${bookingCount}`);
    }

    if (Array.isArray(zipCodeLike) && zipCodeLike.length > 0) {
      const zipConditions = zipCodeLike
        .map(() => `recentZIPCode LIKE ?`)
        .join(" OR ");
      conditions.push(`(${zipConditions})`);
      params.push(...zipCodeLike);
    }

    // Combine all conditions into a single WHERE clause
    let whereClause = "";
    if (conditions.length > 0) {
      whereClause = ` WHERE ${conditions.join(" AND ")}`;
    }
    console.log("TTT", whereClause, params);
    
    const query = Prisma.sql([whereClause, ...params]);
   ```
   

Prints
> TTT  WHERE bookingCount > ? AND (recentZIPCode LIKE ? OR recentZIPCode LIKE ?) [ '2', '22%', '21%' ]

and at the query line
> ⨯ TypeError: Expected 4 strings to have 3 values

**Comment by ludralph** [comment, marked as answer]
Hi @Nubebuster 👋 

Thank you for raising this question. Apologies for missing this on Discord. It must have slipped through the cracks. 

Can you try the code below:

```js
const conditions = [];
const params = [];

if (bookingCount !== undefined) {
  conditions.push(Prisma.sql`bookingCount > ${bookingCount}`);
}

if (Array.isArray(zipCodeLike) && zipCodeLike.length > 0) {
  const zipConditions = zipCodeLike.map(zip => Prisma.sql`recentZIPCode LIKE ${zip}`);
  conditions.push(Prisma.sql`(${Prisma.join(zipConditions, ' OR ')})`);
}

// Combine all conditions into a single WHERE clause
let whereClause = Prisma.empty;
if (conditions.length > 0) {
  whereClause = Prisma.sql`WHERE ${Prisma.join(conditions, ' AND ')}`;
}

// Construct the final query using Prisma.sql
const query = Prisma.sql`SELECT * FROM USER ${whereClause}`;
```
Each condition is constructed using `Prisma.sql` to safely inject parameters. The conditions array is joined with `AND` using `Prisma.join` with a string separator ' AND '. The final query is constructed using Prisma.sql, ensuring that the SQL string and the parameters are correctly aligned. This approach ensures that the SQL string and the parameters are correctly aligned, preventing the type mismatch error you encountered.

Do let us know if this resolves the issue for you.

**Reply by NubeBuster** [reply]
ChatGPT and I must have been doing something subtlety wrong with the Prisma.join function. This works like a charm.

**Reply by ludralph** [reply]
I'm glad to hear the issue is resolved for you. 😄
